xDSL: add immutable core IR structures

martin-luecke opened the pull request:
This PR adds an alternative core infrastructure that provides immutability.
Immutability is especially useful for building rewriting systems that support backtracking.

This introduces immutable versions of these core IR constructs:
Operation -> [IOp](https://github.com/xdslproject/xdsl/blob/7987690229bdb16bf528eddf2e8eb78e1b3348db/xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py#L349)
Block -> [IBlock](https://github.com/xdslproject/xdsl/blob/7987690229bdb16bf528eddf2e8eb78e1b3348db/xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py#L213)
Region -> [IRegion](https://github.com/xdslproject/xdsl/blob/7987690229bdb16bf528eddf2e8eb78e1b3348db/xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py#L110)
SSAValue -> [ISSAValue](https://github.com/xdslproject/xdsl/blob/7987690229bdb16bf528eddf2e8eb78e1b3348db/xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py#L55)
OpResult -> [IResult](https://github.com/xdslproject/xdsl/blob/7987690229bdb16bf528eddf2e8eb78e1b3348db/xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py#L78)
BlockArg -> [IBlockArg](https://github.com/xdslproject/xdsl/blob/7987690229bdb16bf528eddf2e8eb78e1b3348db/xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py#L93)

Some implementation decisions are taken to increase the amount of sharing of IR constructs for rewriting. A specific example for this is the splitting off of some fields from `IOp` to [OpData](https://github.com/xdslproject/xdsl/blob/7987690229bdb16bf528eddf2e8eb78e1b3348db/xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py#L336)


The immutable IR is designed to enable Python pattern matching on it in a non verbose, pythonic way. 
For this reason some of the names of IR constructs are shortened.

math-fehr commented on xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py:
```diff
@@ -0,0 +1,550 @@
+        """Creates a new immutable block."""
+
+        # Type Guards:
+        def is_iblock_arg_seq(
```
We could think of moving this function (if we can make it generic) to `utils`.

math-fehr commented on xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py:
```diff
@@ -0,0 +1,550 @@
+        for arg in self.args:
+            object.__setattr__(arg, "block", self)
+
+    def __init__(self, args: Sequence[Attribute] | Sequence[IBlockArg],
```
What is the case where we give a list of `IBlockArg`?

martin-luecke replied:
This is intended for during rewriting in the case that existing IBlockArgs should be reused. I am not quite 100% sure reuse still works in this design, as the IBlockArg holds a reference to its defining IBlock.
I am not quite sure whether some rewrites are expressed easier when the Blockargs can be created independently from the block first. So for now I am removing this.

math-fehr commented on xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py:
```diff
@@ -0,0 +1,550 @@
+        if block_map is None:
+            block_map = {}
+
+        args: List[IBlockArg] = []
```
The way I see it, you are putting dummy `IBlockArg`, so users can point to it when creating the operations in `Block`.
Instead, could we create the `IBlock` with the arguments, and then add the operations?
That way, we don't need this "hack" of having `IBlockArg` with a `None` block.

martin-luecke replied:
Yes, thank you. I restructured it

math-fehr commented on xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py:
```diff
@@ -0,0 +1,550 @@
+            if operand in value_mapping:
+                mutable_operands.append(value_mapping[operand])
+            else:
+                print(f"ERROR: op {self.name} uses SSAValue before definition")
```
We should fix this at some point, but I believe this doesn't work when blocks are not in order (which technically could happen)

math-fehr replied:
Also, we should probably return an exception rather than just printing something.

martin-luecke replied:
> We should fix this at some point, but I believe this doesn't work when blocks are not in order (which technically could happen)

Ah, you mean when an op uses the result from an op in a block below, but which dominates the current block?

martin-luecke replied:
> Also, we should probably return an exception rather than just printing something.

For debugging it is incredibly useful to still form a complete (although with ill formed references) program and print it so the error can be seen in context. But I agree, we should come up with a more principled way than this.

The problem is, the immutable IR can not be printed on its own, without converting to mutable first.

webmiche replied:
Maybe there exists a way to combine this with our inlined exception prints? (However, that will first have to be readded to the printer, I think).

martin-luecke replied:
Is there progress or a pathway to achieve the behavior you are after or should I leave it like this?

I would like to report an error here, but still continue with the process such that the error can be examined in context. 
I could just move this to throw an exception, but this is so so valuable to me. I would then prefer to have a `DEVMODE` flag or something to enable continuing even though there were errors.

webmiche replied:
I am fine with leaving as is. @math-fehr ?

math-fehr commented on xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py:
```diff
@@ -0,0 +1,550 @@
+            if successor in block_mapping:
+                mutable_successors.append(block_mapping[successor])
+            else:
+                raise Exception("Block used before definition")
```
```suggestion
                raise Exception("Invalid IR: Block is not defined in the current region")
```

math-fehr replied:
We should maybe create an `InvalidIRException` or something.

martin-luecke replied:
Yes, I agree. Maybe something like `ScopingException`

webmiche replied:
I think `InvalidIREception` sounds quite a bit more general, so i would personally go with that...

math-fehr replied:
I think we should have a hierarchy of them ;)

math-fehr commented on xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py:
```diff
@@ -0,0 +1,550 @@
+            if successor in block_map:
+                successors.append(block_map[successor])
+            else:
+                # TODO: I think this is not right, build tests with successors
```
From my understanding, this should never run, since (dummy) blocks should have been created at that point.

martin-luecke replied:
I think the only possible case is when `from_mutable` is called on an op in a region without calling it on the parent op. I replaced the else branch here with an Exception.

webmiche replied:
I don't see the exception. Did you forget to push at some point?

martin-luecke replied:
Indeed, I did now

math-fehr commented on xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py:
```diff
@@ -0,0 +1,550 @@
+        return self.attributes[name]
+
+    def get_attributes_copy(self) -> dict[str, Attribute]:
+        return self.attributes.copy()
```
Can we add a line break at the end?

martin-luecke replied:
:+1:

math-fehr commented on xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py:
```diff
@@ -0,0 +1,550 @@
+    """
+    name: str
+    op_type: type[Operation]
+    attributes: dict[str, Attribute]
```
Is there a reason why this is mutable?
Should we have an `IDict`?

martin-luecke replied:
Right, this should not be mutable. For Python there already exists a `frozendict`, which we could pick up as a dependency. 
I opted to reimplement the `frozenlist` to achieve nicer pattern matching. I have not experimented with how pattern matching looks like with `frozendict`

martin-luecke replied:
I moved the implementation to `frozendict` but sadly it seems it does not properly support type annotations.

math-fehr commented on tests/test_ir.py:
```diff
@@ -232,16 +233,37 @@ def test_is_structurally_equivalent_incompatible_ir_nodes():
+            program.ops[0].regions[0].blocks[1]) == False
+
+
+@pytest.mark.parametrize("program_str", [(program_region), (program_region_2),
```
I would probably create a new file to test `immutable_ir`, so our files are well separated.
I would probably create a new folder, or a new `immutable_ir` file.

martin-luecke replied:
I did think so myself, but opted to include it here so as to not duplicate all of the IR strings in this file. If you prefer this in a separate file with duplicated code (or can we import them maybe?), I can do that

math-fehr replied:
Oh I see, I didn't realize at the moment that we need the same files.
I'm fine to either keep this as it is, or create a new file with the examples, and then separating the files.

math-fehr commented on tests/test_ir.py:
```diff
@@ -232,16 +233,37 @@ def test_is_structurally_equivalent_incompatible_ir_nodes():
 
     parser = Parser(ctx, program_func)
-    program: ModuleOp = parser.parse_op()
+    program: Operation = parser.parse_op()
```
Why was this part changed?

martin-luecke replied:
This fixes a typing error: `parser.parse_op` returns an `Operation`. I missed this in the previous PR and fixed it here

math-fehr reviewed (commented):
Nice! I added a few comments.
Should we add this file directly in `xdsl/composable_rewriting/immutable_ir.py` directly, or is there other files you plan to add that makes the PR structure better?

math-fehr commented:
Also, since there is only a single `match` that we could replace with an `isinstance` in this file, we should maybe drop it to use yapf for formatting?

georgebisbas reviewed (changes requested):
I will add a review at a later stage, as there are already a few comments to address from @math-fehr .

martin-luecke commented on xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py:
```diff
@@ -0,0 +1,550 @@
+
+
+_T = TypeVar('_T')
+class IList(List[_T]):
```
We could have a new file `immutable_utils.py` or `immutable_list.py` under xdsl.utils. That would be a better place, I think. Do you like either of these?

math-fehr replied:
I'm fine with `immutable_list.py`!

martin-luecke commented on xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py:
```diff
@@ -0,0 +1,550 @@
+
+
+_T = TypeVar('_T')
+class IList(List[_T]):
```
It seems to me that inheriting from `List` is the only way to enable proper pattern matching on custom lists. At least I did not find another way of doing it. e.g. using the builtin frozenlist does not support it.

I overwrote `__set_item__`, `__add__`, `__iadd__`, and `__eq__` to counter these issues.

martin-luecke commented on xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py:
```diff
@@ -0,0 +1,550 @@
+        return self is __o
+
+    @property
+    def block(self) -> Optional[IBlock]:
```
:+1:

martin-luecke commented on xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py:
```diff
@@ -0,0 +1,550 @@
+        return None
+
+    @property
+    def ops(self) -> IList[IOp]:
```
fixed, I aligned it with how ir.py behaves

martin-luecke commented on xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py:
```diff
@@ -0,0 +1,550 @@
+    index: int
+
+    def __hash__(self) -> int:
+        return hash(id(self))
```
Fixed, this now behaves like `IOpResult` :+1:

martin-luecke commented on xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py:
```diff
@@ -0,0 +1,550 @@
+        return self is __o
+
+    def __repr__(self) -> str:
+        return "block of" + str(len(self.ops)) + " with args: " + str(
```
:+1:

webmiche replied:
This was not applied, right?

martin-luecke replied:
Github autohiding unresolved stuff got me here. I commited the suggestion

martin-luecke commented on xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py:
```diff
@@ -0,0 +1,550 @@
+    def get_mutable_copy(
+            self,
+            value_mapping: Optional[dict[ISSAValue, SSAValue]] = None,
+            block_mapping: Optional[dict[IBlock, Block]] = None) -> Region:
```
:+1:

martin-luecke commented on xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py:
```diff
@@ -0,0 +1,550 @@
+
+        return region
+
+    def get_mutable_copy(
```
I think `get_mutable_copy` clearly states that we create a copy and do not change anything about the existing mutable Ops/Regions/Blocks. So I am not sure about this change

math-fehr replied:
I'm fine with `to_mutable_copy` as well, to make extra sure that people understand this is a copy.
But to be honest, this was only a nit comment, I'm not feeling that strongly about it, so the current name is fine for me.

martin-luecke replied:
I changed it to `to_mutable` :+1:

martin-luecke commented on xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py:
```diff
@@ -0,0 +1,550 @@
+        cls,
+        blocks: Sequence[Block],
+        value_map: Optional[dict[SSAValue, ISSAValue]] = None,
+        block_map: Optional[dict[Block, IBlock]] = None,
```
:+1:

martin-luecke commented on xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py:
```diff
@@ -0,0 +1,550 @@
+
+
+_T = TypeVar('_T')
+class IList(List[_T]):
```
see above

martin-luecke commented on xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py:
```diff
@@ -0,0 +1,550 @@
+                except ValueError:
+                    continue
+                # replace dummy successor with actual successor
+                object.__setattr__(op, "successors", IList(op.successors[:dummy_index]+[imm_block]+op.successors[dummy_index+1:]))
```
:+1:

martin-luecke commented:
I changed the immutable structures to a more homogeneous naming scheme:
```
Operation -> IOperation
Block -> IBlock
Region -> IRegion
SSAValue -> ISSAValue
OpResult -> IOpResult
BlockArg -> IBlockArg
```
- I removed the only `match` statement so yapf could do its work. (We should really look at finding a replacement for it)
- addressed your other comments

webmiche commented on xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py:
```diff
@@ -0,0 +1,607 @@
+        """Creates a new immutable block."""
+
+        # Type Guards:
+        def is_iblock_arg_seq(
```
I don't quite like the naming here. Why is it called `is...` and then returns the list? I understand that this is convenient to use, but I would personally name it something like `to_iblock_arg_seq` and then just return and empty list (or None) if the initial list was empty.

Considering that this is not part of our public API, this is not a very strong opinion. (though if we make it part of a utils file as Mathieu suggested, I think this should be renamed.)

math-fehr replied:
I think you got confused there, this returns a `TypeGuard` which is just a boolean ;)

webmiche commented on xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py:
```diff
@@ -0,0 +1,607 @@
+                return False
+            return all([isinstance(elem, IBlockArg) for elem in list])
+
+        def is_type_seq(list: Sequence[Any]) -> TypeGuard[Sequence[Attribute]]:
```
why does this behave differently than the other one for an empty list? Feels like not very intuitive design.

martin-luecke replied:
I agree and simplified this

webmiche commented on xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py:
```diff
@@ -0,0 +1,607 @@
+        return mutable_block
+
+
+def get_immutable_copy(op: Operation) -> IOperation:
```
Shouldn't this be part of the (mutable) Operation class? Feels like it would make more sense there.

martin-luecke replied:
I agree this makes more sense there.
The problem is we then create a circular import dependency, which I am not sure how to best break.  Do you have an idea around this?

webmiche replied:
The classic solution is inline imports, I think, but maybe leave as is for now and refactor later :+1:

webmiche commented on xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py:
```diff
@@ -0,0 +1,607 @@
+
+
+@dataclass(frozen=True)
+class OpData:
```
Is this class "only" an optimization? Or in other words, could this PR exist without it?

martin-luecke replied:
Yes indeed, I removed this for now. We can add this with or after the rewriting PR.

webmiche commented on xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py:
```diff
@@ -0,0 +1,607 @@
+            successors=mutable_successors,
+            regions=mutable_regions)
+
+        for idx, result in enumerate(self.results):
```
Maybe add a comment that this will have an effect when this is called as part of a different conversion. I was personally a bit confused why this is required at first.

martin-luecke replied:
:+1:

webmiche reviewed (changes requested):
LGTM overall. I think one general design decision that I want to point out/discuss is that you are creating the immutable IR with a `from_mutable` and `to_mutable` method. A different way of doing this would be to include the `from_mutable` method as a `to_immutable` method in the mutable classes. I feel that your current design makes the immutable IR feel a bit less central than the mutable one. That can definitely be a fine choice, if we decide to for example only really use it during rewriting, but I still feel it should be a conscious decision. I am also not quite sure whether this wouldn't be technically challenging, in particular around circular imports.

math-fehr commented on xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py:
```diff
@@ -0,0 +1,607 @@
+class ISSAValue(ABC):
+    """
+    Represents an immutable SSA variable. An immutable SSA variable is either an operation result
+    or a basic block argument.
```
I would disagree, especially since this is the comment people will directly see on their IDE.
Knowing that there are only two cases here is pretty valuable IMO.

math-fehr commented on xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py:
```diff
@@ -0,0 +1,607 @@
+class IOpResult(ISSAValue):
+    """Represents an immutable SSA variable defined by an operation result."""
+    op: IOperation
+    result_index: int
```
I just looked at it, and it seems I made that mistake on `OpResult` as well.
I would rename this to `index`, and change the same thing on the mutable IR.

martin-luecke replied:
For `OpResult` in ir.py this is now tracked here: https://github.com/xdslproject/xdsl/issues/559

math-fehr commented on xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py:
```diff
@@ -0,0 +1,607 @@
+    def __init__(self, blocks: Sequence[IBlock]):
+        """Creates a new immutable region from a sequence of immutable blocks."""
+
+        object.__setattr__(self, "blocks", IList(blocks))
```
Yeah, this is the only way to instantiate a frozen dataclass AFAIK

martin-luecke commented on xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py:
```diff
@@ -0,0 +1,607 @@
+    def _remove_user(self, op: IOperation):
+        if op not in self.users:
+            raise Exception(
+                f"Trying to remove a user ({op.name}) that is not an actual user of this value!"
```
What do you mean with "work in this case"? 

`_remove_user` is internal to the immutable IR and when a user which was not really a user is removed something suspicious is going on so I think an exception for this is valuable.

math-fehr replied:
Okay, then let's leave it as it is.
My belief is that `Exception` should be triggered when this is a user error, while bugs in the code should be triggered by `assert`. But since we don't really follow that guideline yet, I'm fine leaving it!

martin-luecke commented on xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py:
```diff
@@ -0,0 +1,607 @@
+
+
+@dataclass(frozen=True)
+class IRegion:
```
I agree but let's keep this PR to the core immutable IR

martin-luecke commented on xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py:
```diff
@@ -0,0 +1,607 @@
+            for block in blocks
+        ]
+
+        assert (blocks[0].parent is not None)
```
I often use asserts to indicate the typing to the type checker, because `blocks[0].parent` is `Region | None` here.
I changed this to an InvalidIRException, which makes much more sense.

martin-luecke commented on xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py:
```diff
@@ -0,0 +1,607 @@
+                for op in block.ops:
+                    try:
+                        dummy_index = op.successors.index(dummy_block)
+                    except ValueError:
```
I went for this solution to avoid the duplicated lookup of:
```
if dummy_block in op.successors:
    dummy_index = op.successors.index(dummy_block)
    ...
```
But not sure about the overhead of throwing and catching the Exception anyway, so I changed it to this now.

webmiche commented on tests/rewriting/composable_rewriting/immutable_ir/test_immutable_ir.py:
```diff
@@ -0,0 +1,83 @@
+    immutable_program = get_immutable_copy(program)
+    mutable_program = immutable_program.to_mutable()
+
+    assert program.is_structurally_equivalent(mutable_program)
```
```suggestion
    assert program.is_structurally_equivalent(mutable_program)
    
```

webmiche reviewed (changes requested):
There is some minor comments left, mainly suggestions. Afterward, its probably good to go.

math-fehr reviewed (approved):
Nice! I added some really nit comments, but otherwise I'm fine merging it!
I'll just check with @PapyChacal if we need to fix something first in the notebooks

webmiche reviewed (approved):
Cool stuff! Let's wait for @PapyChacal s approval about the jupyter situation and then we can merge.

PapyChacal reviewed (approved):
This doesn't prevent the current notebooks to work in JupyterLite :+1:

math-fehr commented:
Okay, then that's good for me to merge.
@georgebisbas, could you look into it?

georgebisbas commented on tests/rewriting/composable_rewriting/immutable_ir/test_immutable_ir.py:
```diff
@@ -0,0 +1,83 @@
+import pytest
```
nit: Are we sure we want these tests in this subfolder hierarchy?

martin-luecke replied:
I am mirroring the hierarchy of the project in the tests. We could also go for a flatter hierarchy if we prefer. 
I have no strong opinion on this. Only a desire to complete this PR

math-fehr replied:
What other hierarchy would you be proposing?
I feel that this one is fine. We could remove maybe the last folder, but otherwise I don't see how this would be a problem?

georgebisbas replied:
good then!

martin-luecke commented on xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py:
```diff
@@ -0,0 +1,607 @@
+@dataclass(frozen=True)
+class ISSAValue(ABC):
+    """
+    Represents an immutable SSA variable. An immutable SSA variable is either an operation result
```
For these comments I mirrored what we have in `ir.py`, where the names of the other classes are not mentioned in the comments.
 
Do you think this makes the structure easier to understand? I would be happy to add this, but don't really think its necessary with the two classes inheriting from `ISSAValue` directly beneath.

Do you want me to do this for the descriptions of all classes in this file? 
Also if we want this here we should probably add it to `ir.py` as well, no?

math-fehr replied:
I think this could help newcomers, and we should probably also use this in our files.
We could add this in a separate PR, to clean them both at the same time, what do you think?

georgebisbas replied:
Happy for that, but please open an issue so that we do not forget. Sometimes we tend to forget doing these additional stuff

martin-luecke replied:
This is now tracked here: https://github.com/xdslproject/xdsl/issues/560

georgebisbas reviewed (commented):
Can the immutable ir actually support any of the current tests in the repo? Does it have any limitations?

georgebisbas reviewed (commented):
I see that here: https://app.codecov.io/gh/xdslproject/xdsl/pull/334/blob/xdsl/utils/immutable_list.py
Most of the functions are not tested?

[ 9:47 AM](https://xdsl.zulipchat.com/#narrow/pm-with/574108,574174-pm/near/342259086.2E01)
Right?
They are like not used anywhere?

martin-luecke commented:
@georgebisbas Good catch, I added tests for immutable_list.py [here](https://github.com/xdslproject/xdsl/blob/immutable_ir/tests/test_immutable_list.py).

georgebisbas reviewed (approved):
Thanks for these additons @martin-luecke !
